README.md

## Purpose ##

The `waver` package was developed to provide R code for the calculation of fetch length - the open water distance over which wind can blow along a given direction - and predict wave energy based on the calculated fetch and user-supplied wind and wave monitoring data.

The  `fetch_len` and `fetch_len_multi` functions serve to calculate fetch at single or multiple locations, respectively. These functions accept vector spatial datasets (points, lines and polygons), either in the sf or sp format, and in either projected or unprojected (long/lat) coordinates. Users can specify the bearings for which fetch should be calculated, the number of radii to draw at each bearing (and their angular separation), as well as the maximum fetch length to consider.

These functions are based on existing algorithms published by the USGS[1] and Natural Capital Project[2]. Those tools are implemented in ArcGIS with Python code and use raster land layers instead of vector spatial data.

The `wave_energy` function calculates the wave energy flux from either 
(a) the significant wave height and peak period, as provided by wave models
such as NOAA's [WAVEWATCH III](https://polar.ncep.noaa.gov/waves/index2.shtml); 
or (b) the wind speed, fetch and water depth. The latter calculation is based
on the method presented in the US Army Corps of Engineers Coastal Engineering 
Manual[3].


## References ##

[1] J. Rohweder et al. (2008) "Application of Wind Fetch and Wave Models for Habitat Rehabilitation and Enhancement Projects." USGS report. http://www.umesc.usgs.gov/management/dss/wind_fetch_wave_models.html

[2] InVEST Coastal Vulnerability Model. http://data.naturalcapitalproject.org/invest-releases/documentation/2_2_0/coastal_vulnerability.html

[3] Resio, D.T., Bratos, S.M., and Thompson, E.F. (2003). Meteorology
 and Wave Climate, Chapter II-2. Coastal Engineering Manual.
 US Army Corps of Engineers, Washington DC, 72pp.


## Acknowledgements ##
 
This work was supported by the National Socio-Environmental Synthesis Center (SESYNC) under funding received from the National Science Foundation DBI-1052875.


NEWS.md
waver 0.3.0
==================

* Code updated to use spatial geometry functions from the sf package to replace
imports from the retired rgdal and rgeos packages.

* The package now internally uses sf objects instead of the sp (SpatialPoints/Lines/Polygons)
object format, but sp objects are still accepted as inputs for backwards compatibility.

waver 0.2.1
==================

* Fixed a bug in the default arguments to `fetch_len_multi`


waver 0.2.0
==================

* First release on CRAN


waver_test.R
# File to test waver package

library(waver)

# Read shoreline layer (GSHSS_clip) and point of interest layer (Eco)
GSHHS <- rgdal::readOGR("/nfs/dgill-data/GIS/Contextual data/GSHHS_Shoreline/f", "GSHHS_f_L1")
GSHHS <- as(GSHHS, "SpatialPolygons")
GSHHS_clip <- rgdal::readOGR("/nfs/dgill-data/GIS/Contextual data/GSHHS_Shoreline", "Shoreline_50km_EcoClip_27May15")
GSHHS_clip <- as(GSHHS_clip, 'SpatialPolygons')
Eco <- rgdal::readOGR("/nfs/dgill-data/GIS/Ecological data", "Combined_Eco_data_28May15_landadj")
Eco <- as(Eco, "SpatialPoints")

# Bearings for which to calculate fetch
bearings <- seq(0, 337.5, 22.5)
spread <- seq(-10, 10, 2.5)  # relative sub-bearings to average over
#can also do every 22.5/16 degrees
# Maximum distance
dmax <- 50000

# Fetch for 1st point
system.time(fetch_res <- fetch_len(Eco[1], bearings, GSHHS_clip, dmax, spread))

# Multiple points
system.time(fetchmult <- fetch_len_multi(Eco[1:5], bearings, GSHHS_clip, dmax, spread))

# Projected coords
land <- rgdal::readOGR("/nfs/dgill-data/wind_fetch/Greg/SEFL_LandPolygon_UTM.shp",
                        "SEFL_LandPolygon_UTM")
land <- as(land, "SpatialPolygons")
greg_pts <- rgdal::readOGR("/nfs/dgill-data/wind_fetch/Greg/Shoreline_Points.shp",
                           "Shoreline_Points")
greg_pts <- as(greg_pts, "SpatialPoints")

system.time(fetch_res2 <- fetch_len(greg_pts[2], bearings, land, dmax, spread, TRUE))
system.time(fetchmult2 <- fetch_len_multi(greg_pts[1:5], bearings, land, dmax, spread, TRUE))




tests/testthat.R
library(testthat)
library(waver)

test_check("waver")


tests/testthat/test_fetch_len.R
library(waver)
context("Fetch length")

sf_use_s2(FALSE)

# Set up Spatial objects
longlat <- st_crs("EPSG:4326")
aeqd1 <- st_crs("ESRI:54032")
aeqd2 <- st_crs("+proj=aeqd +lon_0=180 +lat_0=0")

# p1 at (0,0)
p1 <- st_sfc(st_point(c(0, 0)), crs = longlat)
land1 <- st_sfc(waver:::poly_rect(-0.2, 0.25, 0.3, 0.5), crs = longlat)
p1_df <- st_sf(v1 = 2, v2 = "a", p1)
p1_prj <- st_transform(p1, aeqd1)
land1_prj <- st_transform(land1, aeqd1)

# p2 near international date line
p2 <- st_sfc(st_point(c(-179.75, 0)), crs = longlat)
land2 <- st_sfc(list(
    waver:::poly_rect(-179.95, 0.25, -179.45, 0.5),
    waver:::poly_rect(179.7, -0.4, 179.95, 0.2)
), crs = longlat)
p2_prj <- st_transform(p2, aeqd2)
land2_prj <- st_transform(land2, aeqd2)

# p3 on land
p3 <- st_sfc(st_point(c(0, 0.4)), crs = longlat)
p3_prj <- st_transform(p3, aeqd1)
p13 <- c(p1, p3)
p13_prj <- st_transform(p13, aeqd1)

# Multiple points
pts <- c(p1, p2, p3)
lands <- c(land1, land2)
pts_df <- st_sf(v1 = c(2, 5, 6), v2 = c("a", "e", "h"), pts)
lands_df <- st_sf(v3 = c("s", "t", "u"), lands)

# Fetch parameters
bearings <- c(0, 45, 225, 315)
spread <- c(-10, 0, 10)
dmax <- 50000

# Expected results
tol <- 1E-4 # relative tolerance level
fexp1 <- setNames(c(27644, 39094, 50000, 50000), bearings)
fexp1spr <- setNames(c(27926, 40937, 50000, 44610), bearings)
fexp2 <- setNames(c(27644, 39094, 47228, 50000), bearings)
fexp2spr <- setNames(c(27926, 40937, 46005, 44610), bearings)
fexp3 <- setNames(rep(NA, 4), bearings)


test_that("fetch_len correct for single point", {
    expect_equal(fetch_len(p1, bearings, land1, dmax), fexp1, tolerance = tol)
    expect_equal(fetch_len(p1, bearings, land1, dmax, spread),
                 fexp1spr, tolerance = tol)
    expect_equal(fetch_len(p1, bearings, st_cast(land1, "LINESTRING"), dmax),
                 fexp1, tolerance = tol)
    expect_equal(fetch_len(p1_prj, bearings, land1_prj, dmax),
                 fexp1, tolerance = tol)
    expect_equal(fetch_len(p1_prj, bearings, land1_prj, dmax, spread),
                 fexp1spr, tolerance = tol)
    expect_equal(fetch_len(p1_prj, bearings, st_cast(land1_prj, "LINESTRING"), dmax),
                 fexp1, tolerance = tol)
})


test_that("fetch_len correct near international date line", {
    expect_equal(fetch_len(p2, bearings, land2, dmax), fexp2, tolerance = tol)
    expect_equal(fetch_len(p2, bearings, land2, dmax, spread),
                 fexp2spr, tolerance = tol)
    expect_equal(fetch_len(p2_prj, bearings, land2_prj, dmax),
                 fexp2, tolerance = tol)
    expect_equal(fetch_len(p2_prj, bearings, land2_prj, dmax, spread),
                 fexp2spr, tolerance = tol)
})


test_that("fetch_len for point on land returns NA and issues warning", {
    expect_warning(fetch_len(p3, bearings, land1, dmax), "on land")
    expect_warning(fetch_len(p3_prj, bearings, land1_prj, dmax), "on land")
    expect_equal(suppressWarnings(fetch_len(p3, bearings, land1, dmax)), fexp3)
    expect_equal(suppressWarnings(fetch_len(p3_prj, bearings, land1_prj, dmax)), fexp3)
})


test_that("fetch_len_multi matches fetch_len results", {
    expect_equal(suppressWarnings(fetch_len_multi(pts, bearings, lands, dmax)),
                 `rownames<-`(rbind(fexp1, fexp2, fexp3), NULL), tolerance = tol)
    expect_equal(suppressWarnings(fetch_len_multi(pts, bearings, lands, dmax,
                                                  method = "clip")),
                 `rownames<-`(rbind(fexp1, fexp2, fexp3), NULL), tolerance = tol)
    expect_equal(suppressWarnings(fetch_len_multi(pts, bearings, lands, dmax,
                                                  method = "clip")),
                 `rownames<-`(rbind(fexp1, fexp2, fexp3), NULL), tolerance = tol)
    expect_equal(
        suppressWarnings(fetch_len_multi(pts, bearings, lands, dmax, spread)),
        `rownames<-`(rbind(fexp1spr, fexp2spr, fexp3), NULL), tolerance = tol)
    expect_equal(
        suppressWarnings(fetch_len_multi(p13_prj, bearings, land1_prj, dmax)),
        `rownames<-`(rbind(fexp1, fexp3), NULL), tolerance = tol)
})


test_that("fetch_len works with sf objects", {
    expect_equal(fetch_len(p1_df, bearings, lands_df, dmax),
                 fexp1, tolerance = tol)
    expect_equal(
        suppressWarnings(fetch_len_multi(pts_df, bearings, lands_df, dmax)),
        `rownames<-`(rbind(fexp1, fexp2, fexp3), NULL), tolerance = tol)
})


test_that("fetch_len fails on bad inputs", {
    expect_error(fetch_len(0, bearings, land1, dmax), "spatial point")
    expect_error(fetch_len(pts, bearings, land1, dmax), "single point")
    expect_error(fetch_len(p1, bearings, p1, dmax), "lines or polygons")
    expect_error(fetch_len(p1_prj, bearings, land1, dmax), "projections")
    expect_error(fetch_len(p1, bearings, land1_prj, dmax), "projections")
    expect_error(fetch_len(p1_prj, bearings, st_transform(land1, aeqd2),
                           dmax), "projections")
    expect_error(fetch_len(p1, "a", land1, dmax), "bearings")
    expect_error(fetch_len(p1, bearings, land1, dmax, "a"), "spread")
    expect_error(fetch_len(p1, bearings, land1, 0), "dmax")
    expect_error(fetch_len(p1, bearings, land1, c(1,2)), "dmax")
})


R/sp_utils.R
#### Some utility functions to create / manipulate spatial objects (not exported)

# Create a polygon (simple feature) corresponding to rectangle with given coords
poly_rect <- function(xmin, ymin, xmax, ymax) {
    st_polygon(list(cbind(c(rep(xmin, 2), rep(xmax, 2), xmin),
                          c(ymin, rep(ymax, 2), rep(ymin, 2)))))
}

# Create a linestring (simple feature) starting at p, of length len,
#     along given bearing (in degrees)
# Note: This function is meant to be used with a projected CRS
bearing_line <- function(p, bearing, len) {
    l1 <- st_linestring(rbind(
        st_coordinates(p),
        st_coordinates(p) + len * c(sinpi(bearing / 180), cospi(bearing / 180))
    ))
    st_sfc(l1, crs = st_crs(p))
}

# Create clipping rectangle around point p (sfc_POINT object for a single point)
#  to guarantee at least dmax on each side
# dmax either in meters (if longlat coordinates) or in the projection's coordinates
get_clip_rect <- function(p, dmax) {
    if (st_is_longlat(p)) {
        lat_dist <- 111600 # approx. distance (in m) between degrees of latitude
        long <- st_coordinates(p)[1]
        lat <- st_coordinates(p)[2]
        ybuf <- dmax / lat_dist
        xbuf <- ybuf / cospi(abs(lat) / 180)
        # Split clip_rect in two if it would overlap international date line
        if (long - xbuf < -180) {
            westr <- poly_rect(-180, lat - ybuf, long + xbuf, lat + ybuf)
            eastr <- poly_rect(long - xbuf + 360, lat - ybuf, 180, lat + ybuf)
            suppressMessages(
                clip_rect <- st_union(st_sfc(westr, eastr, crs = st_crs(p)))
            )
        } else if(long + xbuf > 180) {
            westr <- poly_rect(-180, lat - ybuf, long + xbuf - 360, lat + ybuf)
            eastr <- poly_rect(long - xbuf, lat - ybuf, 180, lat + ybuf)
            suppressMessages(
                clip_rect <- st_union(st_sfc(westr, eastr, crs = st_crs(p)))
            )
        } else {
            r1 <- poly_rect(long - xbuf, lat - ybuf, long + xbuf, lat + ybuf)
            clip_rect <- st_sfc(r1, crs = st_crs(p))
        }
    } else {
        x <- st_coordinates(p)[1]
        y <- st_coordinates(p)[2]
        r1 <- poly_rect(x - dmax, y - dmax, x + dmax, y + dmax)
        clip_rect <- st_sfc(r1, crs = st_crs(p))
    }
    clip_rect
}



R/wave_energy.R
g_acc <- 9.81 # gravitational acceleration (m/s^2)

# Wave energy constant (using density of water at 20 C i.e. 998 kg/m^3)
we_const <- 0.998 * g_acc^2 / (64 * pi)


#' Calculate the wave energy flux
#'
#' Calculates the wave energy flux (power per meter of wave crest) given
#' either (1) the significant wave height and peak period or (2) the wind speed
#' at 10m, fetch length and (optionally) water depth.
#'
#' Given the significant height (\emph{H}) and peak period (\emph{T}),
#' the wave energy flux is calculated as: \deqn{\frac{\rho g^2}{64 \pi} H^2 T},
#' where \eqn{\rho} is the density of water (998 kg/m^3) and \emph{g} is the
#' acceleration of gravity (9.81 m/s^2).
#'
#' If both \code{height} and \code{period} are missing, they are estimated from
#' on the wind speed at 10m (\eqn{U_{10}}) and the fetch length (\emph{F}) as
#' described in Resio et al. (2003):
#' \deqn{{U_f}^2 = 0.001 (1.1 + 0.035 U_{10}) {U_{10}}^2} (friction velocity)
#' \deqn{\frac{g H}{{U_f}^2} = \min (0.0413 \sqrt{\frac{g F}{{U_f}^2}}, 211.5)}
#' \deqn{\frac{g T}{U_f} = \min (0.651 (\frac{g F}{{U_f}^2})^{1/3}, 239.8)}
#' If the depth (\emph{d}) is specified, it imposes a limit on the peak period:
#' \deqn{T_{max} = 9.78 \sqrt{\frac{d}{g}}} (in seconds)
#'
#' @param height Significant wave height, in meters.
#' @param period Peak wave period, in seconds.
#' @param wind Wind speed at 10m, in m/s.
#' @param fetch Fetch length, in meters.
#' @param depth Water depth, in meters.
#' @return Wave energy flux, in kW/m.
#' @references Resio, D.T., Bratos, S.M., and Thompson, E.F. (2003). Meteorology
#'  and Wave Climate, Chapter II-2. Coastal Engineering Manual.
#'  US Army Corps of Engineers, Washington DC, 72pp.
#' @examples
#'  # With height and period arguments
#'  wave_energy(8, 1)
#'
#'  # With wind, fetch and depth arguments (must be named)
#'  wave_energy(wind = 12, fetch = 15000, depth = 10)
#' @export
wave_energy <- function(height = NA, period = NA,
                        wind = NA, fetch = NA, depth = NA) {
    if (all(is.na(height)) && all(is.na(period))) {
        # Friction velocity squared
        uf2 <- 0.001 * (1.1 + 0.035 * wind) * wind^2
        # Non-dimensional fetch, height and period
        fetch_nd <- fetch * g_acc / uf2
        height_nd <- pmin(0.0413 * sqrt(fetch_nd), 211.5)
        period_nd <- pmin(0.651 * fetch_nd^(1/3), 239.8)
        # Calculate height and period (latter with depth limitation, if available)
        #  then wave energy
        height <- height_nd * uf2 / g_acc
        period <- period_nd * sqrt(uf2) / g_acc
        if (any(!is.na(depth)))
            period <- pmin(period, 9.78 * sqrt(depth / g_acc))
    } else {
        if (any(!is.na(wind)) || any(!is.na(fetch)) || any(!is.na(depth)))
            warning("height or period provided; ignoring other arguments to wave_energy.")
    }
    we_const * height^2 * period
}


R/fetch_len_proj.R

#' Calculate the fetch length around a point (projected coordinates)
#'
#' Given a point, a shoreline layer and a vector of wind directions (bearings),
#' \code{fetch_len_proj} calculates the distance from point to shore for each bearing.
#' This version of the function assumes the point and shoreline share the same projection.
#'
#' The fetch length (or fetch) is the distance of open water over which the wind
#' can blow in a specific direction. Note that bearings represent the direction
#' from where the wind originates.
#'
#' The optional \code{spread} argument defines relative directions that are
#' added to each main bearing to produce a set of sub-bearings. The fetch lengths
#' calculated for each sub-bearing are averaged with weights proportional to
#' \code{cos(spread)}. By default, \code{spread = 0} and fetch length is
#' calculated for the main bearings only.
#'
#' If the shoreline layer is given as SpatialPolygons*, the function verifies
#' that the input point is outside all polygons (i.e. in water). If this is
#' not the case, it issues a warning and returns a vector of \code{NA}.
#'
#' @param p SpatialPoints* object of length 1 (single point).
#' @param bearings Vector of bearings, in degrees.
#' @param shoreline SpatialLines* or SpatialPolygons* object representing the
#'  shoreline, in the same projection as \code{p}.
#' @param dmax Maximum value of fetch length, returned if there is no land
#'  within a distance of \code{dmax} from a given bearing. This should be
#'  in the same units as the coordinates of \code{p} and \code{shoreline}.
#' @param spread Vector of relative bearings (in degrees) for which
#'  to calculate fetch around each main bearing (see details).
#' @return A named vector representing the fetch length for each direction
#'  given in \code{bearings}.
#' @seealso \code{\link{fetch_len}} to work in geographic (unprojected)
#'  coordinates.
#' @export
fetch_len_proj <- function(p, bearings, shoreline, dmax, spread = 0) {
    # Check inputs
    if (!is(p, "SpatialPoints") || length(p) != 1) {
       stop("p must be a SpatialPoints* object of length 1.")
    }
    if (!(is(shoreline, "SpatialLines") || is(shoreline, "SpatialPolygons"))) {
        stop("shoreline must be a SpatialLines* or SpatialPolygons* object.")
    }
    if (proj4string(p) != proj4string(shoreline)) {
        stop("projections of p and shoreline do not match.")
    }
    if (!is.vector(bearings, "numeric")) stop("bearings must be a numeric vector.")
    if (!is.vector(spread, "numeric")) stop("spread must be a numeric vector.")
    if (!is.vector(dmax, "numeric") || length(dmax) != 1) {
        stop("dmax must be a single number.")
    }

    # If shoreline is a polygons (land) layer, check that point is not on land
    if (is(shoreline, "SpatialPolygons")) {
        # Note: 'as' only to remove DataFrame part of Spatial objects
        in_water <- is.na(over(as(p, "SpatialPoints"),
                               as(shoreline, "SpatialPolygons")))
        if(!in_water) {
            warning("point on land, returning NA")
            return(rep(NA, length(bearings)))
        }
    }

    # Clip shoreline layer to a rectangle around point
    # to guarantee at least dmax on each side
    clip_rect <- get_clip_rect_proj(p, dmax)
    shore_clip <- rgeos::gIntersection(shoreline, clip_rect, byid = TRUE)
    # If no land within rectangle, return dmax for all bearings
    if (is.null(shore_clip) || is(shore_clip, "SpatialPoints")) {
        return(rep(dmax, length(bearings)))
    }

    # Convert any polygons to lines to get line-line intersections later
    if (is(shore_clip, "SpatialPolygons")) {
        shore_clip <- as(shore_clip, "SpatialLines")
    } else if (is(shore_clip, "SpatialCollections")) {
        if (!is.null(shore_clip@polyobj)) {
            shore_clip <- rbind(as(shore_clip@polyobj, "SpatialLines"),
                                shore_clip@lineobj)
        } else if (!is.null(shore_clip@lineobj)) {
            shore_clip <- shore_clip@lineobj
        } else {  # no land in buffer
            return(rep(dmax, length(bearings)))
        }
    }

    if (all(spread == 0)) {
        # if no sub-bearings, just return distance to shore for each bearing
        fetch_res <- vapply(bearings,
                            function(b) dist_shore_proj(p, shore_clip, b, dmax), 0)
    } else {
        # calculate the distance to shore for each sub-bearing
        bear_mat <- outer(bearings, spread, "+")
        dists <- vapply(bear_mat,
                        function(b) dist_shore_proj(p, shore_clip, b, dmax), 0)
        dim(dists) <- dim(bear_mat)
        # return weighted means of the sub-bearing fetch values
        #  with weights proportional to the cosine (relative to their main bearing)
        weights <- cospi(spread / 180)
        weights <- weights / sum(weights)
        fetch_res <- as.vector(dists %*% weights)
    }

    names(fetch_res) <- as.character(bearings)
    fetch_res
}


# Helper functions below are not exported by the package

# Create clipping rectangle around point p
#   to guarantee at least dmax on each side
get_clip_rect_proj <- function(p, dmax) {
    x <- coordinates(p)[1]
    y <- coordinates(p)[2]
    r1 <- poly_rect(x - dmax, y - dmax, x + dmax, y + dmax)
    SpatialPolygons(list(Polygons(list(r1), ID = 1)),
                    proj4string = CRS(proj4string(p)))
}

# Returns the distance from point p to shoreline
# following bearing bear and up to distance dmax
dist_shore_proj <- function(p, shoreline, bear, dmax) {
    # Draw line of length dmax with given start point and bearing
    bline <- bearing_line(p, bear, dmax)
    # Return (minimum) distance from p1 to intersection of geo_line and shoreline
    # If no intersection, fetch is dmax
    land_int <- rgeos::gIntersection(bline, shoreline)
    if (is.null(land_int)) {
        dmax
    } else {
        rgeos::gDistance(p, land_int)
    }
}


R/fetch_len.R

#' Calculate the fetch length around a point
#'
#' Given a point, a shoreline layer and a vector of wind directions (bearings),
#' \code{fetch_len} calculates the distance from point to shore for each bearing.
#'
#' The fetch length (or fetch) is the distance of open water over which the wind
#' can blow in a specific direction. Note that bearings represent the direction
#' from where the wind originates.
#'
#' The optional \code{spread} argument defines relative directions that are
#' added to each main bearing to produce a set of sub-bearings. The fetch lengths
#' calculated for each sub-bearing are averaged with weights proportional to
#' \code{cos(spread)}. By default, \code{spread = 0} and fetch length is
#' calculated for the main bearings only.
#'
#' The input data can be in either geographic (long, lat) or projected coordinates,
#' but \code{p} and \code{shoreline} must share the same coordinate system. Distances
#' are calculated using the \code{\link[sf]{st_distance}} function from the sf package
#' and expressed in the units of the coordinate system used, or in meters if using
#' geographic coordinates. For geographic coordinates, we recommend setting
#' \code{sf_use_s2(FALSE)}, which results in \code{st_distance} using the ellipsoid
#' distance calculation (requires the lwgeom package), instead of the less precise
#' spherical distance calculation. For projected coordinates, the Euclidean distance
#' is calculated.
#'
#' If the shoreline layer is composed of polygons rather than lines, the function
#' verifies that the input point is outside all polygons (i.e. in water). If this is
#' not the case, it issues a warning and returns a vector of \code{NA}.
#'
#' @param p Simple feature (sf or sfc) object representing a single point.
#' @param bearings Vector of bearings, in degrees.
#' @param shoreline Simple feature (sf or sfc) object representing the
#'  shoreline, in either line or polygon format.
#' @param dmax Maximum value of fetch length, returned if there is no land
#'  within a distance of \code{dmax} from a given bearing.
#' @param spread Vector of relative bearings (in degrees) for which
#'  to calculate fetch around each main bearing (see details).
#' @param projected Deprecated argument, kept for backwards compatibility.
#' @param check_inputs Should the validity of inputs be checked? It is
#' recommended to keep this TRUE, unless this function is called repeatedly from
#' another function that already checks inputs.
#' @return A named vector representing the fetch length for each direction
#'  given in \code{bearings}.
#' @examples
#'  pt <- st_sfc(st_point(c(0, 0)), crs = st_crs(4326))
#'  # Shoreline is a rectangle from (-0.2, 0.25) to (0.3, 0.5)
#'  rect <- st_polygon(list(cbind(c(rep(-0.2, 2), rep(0.3, 2), -0.2),
#'                                c(0.25, rep(0.3, 2), rep(0.25, 2)))))
#'  land <- st_sfc(rect, crs = st_crs(4326))
#'  fetch_len(pt, bearings = c(0, 45, 225, 315), land,
#'            dmax = 50000, spread = c(-10, 0, 10))
#' @seealso \code{\link{fetch_len_multi}} for an efficient alternative when
#'  computing fetch length for multiple points.
#' @export
fetch_len <- function(p, bearings, shoreline, dmax, spread = 0,
                      projected = FALSE, check_inputs = TRUE) {
    # Convert sp objects to sf format
    if(is(p, "SpatialPointsDataFrame")) p <- st_as_sf(p)
    if(is(p, "SpatialPoints")) p <- st_as_sfc(p)
    if(is(shoreline, "SpatialLinesDataFrame") || is(shoreline, "SpatialPolygonsDataFrame")) {
        shoreline <- st_as_sf(shoreline)
    }
    if(is(shoreline, "SpatialLines") || is(shoreline, "SpatialPolygons")) {
        shoreline <- st_as_sfc(shoreline)
    }
    # Extract geometry columns if p or shoreline are sf objects
    if(is(p, "sf")) p <- st_geometry(p)
    if(is(shoreline, "sf")) shoreline <- st_geometry(shoreline)

    if (check_inputs) {
        if (!is(p, "sfc_POINT")) stop("p must be a spatial point in sf or sp format.")
        if(length(p) != 1) stop("p must be a single point.")
        if (!(is(shoreline, "sfc_LINESTRING") || is(shoreline, "sfc_MULTILINESTRING") ||
              is(shoreline, "sfc_POLYGON")    || is(shoreline, "sfc_MULTIPOLYGON"))) {
            stop("shoreline must be a spatial object in sf or sp format containing lines or polygons.")
        }
        if (st_crs(p) != st_crs(shoreline)) {
            stop("projections of p and shoreline do not match.")
        }

        if (!is.vector(bearings, "numeric")) stop("bearings must be a numeric vector.")
        if (!is.vector(spread, "numeric")) stop("spread must be a numeric vector.")
        if (!is.vector(dmax, "numeric") || length(dmax) != 1 || dmax <= 0) {
            stop("dmax must be a single number greater than 0.")
        }
    }

    # If shoreline is a polygons (land) layer, check that point is not on land
    if (is(shoreline, "sfc_POLYGON") || is(shoreline, "sfc_MULTIPOLYGON")) {
        suppressMessages(
            in_water <- length(st_intersection(p, shoreline)) == 0
        )
        if(!in_water) {
            warning("point on land, returning NA")
            return(setNames(rep(NA, length(bearings)), bearings))
        }
    }

    # Clip shoreline layer to a rectangle around point
    # to guarantee at least dmax on each side
    clip_rect <- get_clip_rect(p, dmax)
    suppressMessages(
        shore_clip <- st_intersection(shoreline, clip_rect)
    )

    # If no land within rectangle, return dmax for all bearings
    if (length(shore_clip) == 0) {
        return(setNames(rep(dmax, length(bearings)), bearings))
    }

    # Convert to multilinestring (if not already lines)
    #  since line-line intersections are needed later
    shore_clip <- st_cast(shore_clip, "MULTILINESTRING")

    # Calculate fetch
    if (all(spread == 0)) {
        # If no sub-bearings, just return distance to shore for each bearing
        fetch_res <- vapply(bearings,
                   function(b) dist_shore(p, shore_clip, b, dmax), 0)
    } else {
        # Calculate the distance to shore for each sub-bearing
        bear_mat <- outer(bearings, spread, "+")
        dists <- vapply(bear_mat,
                   function(b) dist_shore(p, shore_clip, b, dmax), 0)
        dim(dists) <- dim(bear_mat)
        # Return weighted means of the sub-bearing fetch values
        #  with weights proportional to the cosine (relative to their main bearing)
        weights <- cospi(spread / 180)
        weights <- weights / sum(weights)
        fetch_res <- as.vector(dists %*% weights)
    }
    names(fetch_res) <- as.character(bearings)
    fetch_res
}


#' Calculate the fetch length for multiple points
#'
#' \code{fetch_len_multi} provides two methods to efficiently compute fetch length
#' for multiple points.
#'
#' With \code{method = "btree"} (default), the fetch calculation for each point only uses
#' the geometries within the \code{shoreline} layer that intersect with a rectangular
#' buffer of size \code{dmax} around that point. (The name is based on a previous version
#' of the function that implemented this method using the \code{gBinarySTRtreeQuery} function
#' from the rgeos package.)
#'
#' With \code{method = "clip"}, the \code{shoreline} is clipped to its intersection
#' with a polygon formed by the union of all the individual points' rectangular buffers.
#'
#' In both cases, \code{\link{fetch_len}} is then applied to each point,
#' using only the necessary portion of the shoreline.
#'
#' Generally, the "clip" method will produce the biggest time savings when
#' points are clustered within distances less than \code{dmax} (so their
#' clipping rectangles overlap), whereas the "btree" method will be more
#' efficient when the shoreline is composed of multiple geometrical objects
#' and points are distant from each other.
#'
#' @param pts Simple features (sf or sfc) object containing point data.
#' @param bearings Vector of bearings, in degrees.
#' @param shoreline Simple feature (sf or sfc) object representing the
#'  shoreline, in either line or polygon format.
#' @param dmax Maximum value of fetch length, returned if there is no land
#'  within a distance of \code{dmax} from a given bearing.
#' @param spread Vector of relative bearings (in degrees) for which
#'  to calculate fetch around each main bearing.
#' @param method Whether to use the "btree" (default) or "clip" method.
#'  See below for more details.
#' @param projected Deprecated argument, kept for backwards compatibility.
#' @return A matrix of fetch lengths, with one row by point in \code{pts} and
#'  one column by bearing in \code{bearings}.
#' @seealso \code{\link{fetch_len}} for details on the fetch length computation.
#' @export
fetch_len_multi <- function(pts, bearings, shoreline, dmax,
                     spread = 0, method = "btree", projected = FALSE) {
    # Convert sp objects to sf format
    if(is(pts, "SpatialPointsDataFrame")) pts <- st_as_sf(pts)
    if(is(pts, "SpatialPoints")) pts <- st_as_sfc(pts)
    if(is(shoreline, "SpatialLinesDataFrame") || is(shoreline, "SpatialPolygonsDataFrame")) {
        shoreline <- st_as_sf(shoreline)
    }
    if(is(shoreline, "SpatialLines") || is(shoreline, "SpatialPolygons")) {
        shoreline <- st_as_sfc(shoreline)
    }
    # Extract geometry columns if pts or shoreline are sf objects
    if(is(pts, "sf")) pts <- st_geometry(pts)
    if(is(shoreline, "sf")) shoreline <- st_geometry(shoreline)

    # Check inputs
    match.arg(method, choices = c("btree", "clip"))
    if (!is(pts, "sfc_POINT")) stop("pts must be a spatial point in sf or sp format.")
    if (!(is(shoreline, "sfc_LINESTRING") || is(shoreline, "sfc_MULTILINESTRING") ||
          is(shoreline, "sfc_POLYGON")    || is(shoreline, "sfc_MULTIPOLYGON"))) {
        stop("shoreline must be a spatial object in sf or sp format containing lines or polygons.")
    }
    if (st_crs(pts) != st_crs(shoreline)) {
        stop("projections of p and shoreline do not match.")
    }

    if (!is.vector(bearings, "numeric")) stop("bearings must be a numeric vector.")
    if (!is.vector(spread, "numeric")) stop("spread must be a numeric vector.")
    if (!is.vector(dmax, "numeric") || length(dmax) != 1 || dmax <= 0) {
        stop("dmax must be a single number greater than 0.")
    }

    # Create rectangular buffers around each point
    rect_list <- lapply(1:length(pts),
                        function(i) get_clip_rect(pts[i], dmax))
    rect_buf <- do.call(c, rect_list)

    if (method == "btree") {
        # Generate list of shoreline geometry indices that intersect each rectangle
        suppressMessages(
            btree <- st_intersects(rect_buf, shoreline)
        )
        # Calculate fetch for point at index i using btree
        fetch_i <- function(i) {
            if (length(btree[[i]]) == 0) {
                setNames(rep(dmax, length(bearings)), bearings)
            } else {
                fetch_len(pts[i], bearings, shoreline[btree[[i]]], dmax,
                          spread, check_inputs = FALSE)
            }
        }
        # Calculate fetch for all points and return a (points x bearings) matrix
        fetch_res <- t(vapply(1:length(pts), fetch_i, rep(0, length(bearings))))
    } else { # method == "clip"
        # Clip shoreline to a merged buffer around all points
        suppressMessages({
            rect_buf <- st_union(rect_buf)
            sub_shore <- st_intersection(shoreline, rect_buf)
        })
        fetch_res <- t(
            vapply(1:length(pts),
                   function(i) fetch_len(pts[i], bearings, sub_shore, dmax,
                                         spread, check_inputs = FALSE),
                   rep(0, length(bearings)))
        )
    }
    fetch_res
}


#### Helper functions below are not exported by the package ####

# Returns the distance from point p to shoreline following bearing bear and up to distance dmax
# Uses the geodetic distance if in geographic coordinates,
#   or the Euclidean distance if in projected coordinates
dist_shore <- function(p, shoreline, bear, dmax) {
    if (st_is_longlat(p)) {
        # Draw geodesic line of length dmax with given start point and bearing
        # Line drawn with a point every dmax/500
        geo_line <- geosphere::gcIntermediate(st_coordinates(p),
                                              geosphere::destPoint(st_coordinates(p), bear, dmax),
                                              n = 500, breakAtDateLine = TRUE, addStartEnd = TRUE
        )
        if (is.list(geo_line)) {
            geo_line <- st_multilinestring(geo_line)
        } else {
            geo_line <- st_linestring(geo_line)
        }
        bline <- st_sfc(geo_line, crs = st_crs(shoreline))
    } else {
        # Draw line of length dmax with given start point and bearing
        bline <- bearing_line(p, bear, dmax)
    }
    # Return (minimum) distance from p to intersection of bearing line and shoreline
    # If no intersection, fetch is dmax
    suppressMessages(
        land_int <- st_intersection(bline, shoreline)
    )
    if (length(land_int) == 0) {
        d <- dmax
    } else {
        d <- min(st_distance(p, land_int))
    }
    d
}
